File: frontend/src/api/apiKeyService.ts
import { apiClient } from './client';
import { AxiosResponse } from 'axios';

export interface ApiKey {
  id: number;
  apiKey: string;
  name: string;
  description?: string;
  enabled: boolean;
  expiresAt?: string;
  lastUsedAt?: string;
  createdBy: string;
  createdAt: string;
  updatedAt: string;
}

export interface CreateApiKeyRequest {
  name: string;
  description?: string;
  expiresAt?: string;
}

export interface ApiResponse<T> {
  success: boolean;
  message?: string;
  data?: T;
}

export const apiKeyService = {
  // Create a new API key
  async createApiKey(request: CreateApiKeyRequest): Promise<ApiResponse<ApiKey>> {
    try {
      const response: AxiosResponse<ApiKey> = await apiClient.post('/api-keys', request);
      return {
        success: true,
        data: response.data
      };
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to create API key'
      };
    }
  },

  // Get all API keys
  async getAllApiKeys(): Promise<ApiResponse<ApiKey[]>> {
    try {
      const response: AxiosResponse<ApiKey[]> = await apiClient.get('/api-keys');
      return {
        success: true,
        data: response.data
      };
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to fetch API keys'
      };
    }
  },

  // Get API key by ID
  async getApiKeyById(id: number): Promise<ApiResponse<ApiKey>> {
    try {
      const response: AxiosResponse<ApiKey> = await apiClient.get(`/api-keys/${id}`);
      return {
        success: true,
        data: response.data
      };
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to fetch API key'
      };
    }
  },

  // Delete API key
  async deleteApiKey(id: number): Promise<ApiResponse<void>> {
    try {
      await apiClient.delete(`/api-keys/${id}`);
      return {
        success: true
      };
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to delete API key'
      };
    }
  },

  // Toggle API key status (enable/disable)
  async toggleApiKeyStatus(id: number, enabled: boolean): Promise<ApiResponse<ApiKey>> {
    try {
      const response: AxiosResponse<ApiKey> = await apiClient.post(`/api-keys/${id}/toggle`, { enabled });
      return {
        success: true,
        data: response.data
      };
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to update API key status'
      };
    }
  },

  // Regenerate API key
  async regenerateApiKey(id: number): Promise<ApiResponse<ApiKey>> {
    try {
      const response: AxiosResponse<ApiKey> = await apiClient.post(`/api-keys/${id}/regenerate`);
      return {
        success: true,
        data: response.data
      };
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to regenerate API key'
      };
    }
  }
};

File: Backend/src/main/java/in/osop/messaging_platform/service/ApiKeyService.java
package in.osop.messaging_platform.service;

import in.osop.messaging_platform.dto.ApiKeyDto;
import in.osop.messaging_platform.dto.CreateApiKeyRequest;

import java.util.List;
import java.util.Optional;

public interface ApiKeyService {
    
    /**
     * Create a new API key
     * @param tenantId The tenant ID
     * @param userEmail The email of the user creating the API key
     * @param request The API key creation request
     * @return The created API key
     */
    ApiKeyDto createApiKey(Long tenantId, String userEmail, CreateApiKeyRequest request);
    
    /**
     * Get all API keys for a tenant
     * @param tenantId The tenant ID
     * @return List of API keys
     */
    List<ApiKeyDto> getAllApiKeys(Long tenantId);
    
    /**
     * Get an API key by ID
     * @param tenantId The tenant ID
     * @param apiKeyId The API key ID
     * @return The API key
     */
    ApiKeyDto getApiKeyById(Long tenantId, Long apiKeyId);
    
    /**
     * Delete an API key
     * @param tenantId The tenant ID
     * @param apiKeyId The API key ID
     */
    void deleteApiKey(Long tenantId, Long apiKeyId);
    
    /**
     * Toggle API key status (enable/disable)
     * @param tenantId The tenant ID
     * @param apiKeyId The API key ID
     * @param enabled The new enabled status
     * @return The updated API key
     */
    ApiKeyDto toggleApiKeyStatus(Long tenantId, Long apiKeyId, boolean enabled);
    
    /**
     * Regenerate an API key with a new value
     * @param tenantId The tenant ID
     * @param apiKeyId The API key ID
     * @return The updated API key
     */
    ApiKeyDto regenerateApiKey(Long tenantId, Long apiKeyId);
    
    /**
     * Validate an API key
     * @param apiKey The API key value
     * @return Optional containing the tenant ID if valid, empty if invalid
     */
    Optional<Long> validateApiKey(String apiKey);
    
    /**
     * Update the last used timestamp for an API key
     * @param apiKey The API key value
     */
    void updateApiKeyLastUsed(String apiKey);
    
    /**
     * Cleanup expired API keys
     */
    void cleanupExpiredApiKeys();
}
